driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -0,0 +1,1114 @@
+                operationContext,
+                effectiveWriteConcern.isAcknowledged(),
+                bulkWriteCommand.getOpsAndNsInfo());
+        if (bulkWriteCommandOkResponse == null) {
+            return null;
+        }
+        List<List<BsonDocument>> cursorExhaustBatches = doWithRetriesDisabledForCommand(retryState, "getMore", () ->
+                exhaustBulkWriteCommandOkResponseCursor(connectionSource, connection, bulkWriteCommandOkResponse));
+        ExhaustiveClientBulkWriteCommandOkResponse exhaustiveBulkWriteCommandOkResponse = new ExhaustiveClientBulkWriteCommandOkResponse(
+                bulkWriteCommandOkResponse, cursorExhaustBatches);
+        // `Connection.command` does not throw `MongoWriteConcernException`, so we have to construct it ourselves
+        MongoWriteConcernException writeConcernException = Exceptions.createWriteConcernException(
+                bulkWriteCommandOkResponse, connection.getDescription().getServerAddress());
+        if (writeConcernException != null) {
+            throw new MongoWriteConcernWithResponseException(writeConcernException, exhaustiveBulkWriteCommandOkResponse);
+        }
+        return exhaustiveBulkWriteCommandOkResponse;
+    }
+
+    private <R> R doWithRetriesDisabledForCommand(
+            final RetryState retryState,
+            final String commandDescription,
+            final Supplier<R> actionWithCommand) {
+        Optional<Boolean> originalRetryableCommandFlag = retryState.attachment(AttachmentKeys.retryableCommandFlag());
+        Supplier<String> originalCommandDescriptionSupplier = retryState.attachment(AttachmentKeys.commandDescriptionSupplier())
+                .orElseThrow(Assertions::fail);
+        try {
+            retryState.attach(AttachmentKeys.retryableCommandFlag(), false, true)
+                    .attach(AttachmentKeys.commandDescriptionSupplier(), () -> commandDescription, false);
+            return actionWithCommand.get();
+        } finally {
+            originalRetryableCommandFlag.ifPresent(value -> retryState.attach(AttachmentKeys.retryableCommandFlag(), value, true));
+            retryState.attach(AttachmentKeys.commandDescriptionSupplier(), originalCommandDescriptionSupplier, false);
+        }
+    }
+
+    private List<List<BsonDocument>> exhaustBulkWriteCommandOkResponseCursor(
+            final ConnectionSource connectionSource,
+            final Connection connection,
+            final BsonDocument response) {
+        int serverDefaultCursorBatchSize = 0;
+        try (CommandBatchCursor<BsonDocument> cursor = cursorDocumentToBatchCursor(
+                TimeoutMode.CURSOR_LIFETIME,
+                response,
+                serverDefaultCursorBatchSize,
+                codecRegistry.get(BsonDocument.class),
+                options.getComment().orElse(null),
+                connectionSource,
+                connection)) {
+            cursor.setCloseWithoutTimeoutReset(true);
```

Comment:
`BULK-TODO`: when https://jira.mongodb.org/browse/JAVA-5615 from `main` is merged in `JAVA-4586_bulk-write`, `setCloseWithoutTimeoutReset` should be removed.

Reply:
Done.